sw/cheri/checks/lcd_check.cc

```diff
@@ -150,6 +152,225 @@ static void run_script(Capability<volatile OpenTitanUart> &uart, SpiLcdCap &spi,
+   * @param dc_high Set D/C pin to 1 if true, otherwise 0.
+   *
+   */
+  uint32_t (*gpio_write)(void *handle, bool cs_high, bool dc_high);
+
+  /**
+   * @brief Set the state of the data line output enable.
+   *
+   * @param oe Enable output if true, otherwise disable output (high-Z).
+   *
+   */
+  uint32_t (*oe_write)(void *handle, bool oe);
+
+  /**
+   * @brief Simple cpu delay
+   *
+   * @param ms Time the delay should take in milliseconds.
+   *
+   */
+  void (*timer_delay)(uint32_t ms);
+} LCD_Interface;
+
+uint32_t spi_write(void *handle, uint8_t *data, size_t len) {
+  SpiLcdCap spi = *(SpiLcdCap *)handle;
+  spi->blocking_write(data, len);
+  spi->wait_idle();
+  return len;
+}
+
+uint32_t spi_read(void *handle, uint8_t *data, size_t len) {
+  SpiLcdCap spi = *(SpiLcdCap *)handle;
+  spi->blocking_read(data, len);
+  spi->wait_idle();
+  return len;
+}
+
+uint32_t gpio_write(void *handle, bool cs_high, bool dc_high) {
+  SpiLcdCap spi = *(SpiLcdCap *)handle;
+  set_output_bit(spi, LcdDcPin, dc_high);
+  set_output_bit(spi, LcdCsPin, cs_high);
+  return 0;
+}
+
+uint32_t oe_write(void *handle, bool oe) {
+  SpiLcdCap spi = *(SpiLcdCap *)handle;
+  set_output_bit(spi, LcdWrPin, oe);
+  return 0;
+}
+
+void timer_delay(uint32_t ms) { delay(ms); }
+
+// Determine if an LCD offset of 2 pixels in the narrow dimension and
+// 1 pixel in the wide dimension must be applied for correct function.
+// Involves writing a bundle pixels to the LCD and reading some back
+// from the start of the affected rows to discover the default width.
+//
+// The state of the GM[2:0] config pads of the ST7735 controller within
+// the LCD is the root value we wish to discover, but can only do so
+// by observing side-effects. This function infers GM state from the
+// reset value the of CASET register, which itself must be inferred from
+// pixel buffer behaviour after a reset as it cannot be read directly.
+// The test used is whether the 129th pixel or the 133rd pixel written
+// ends up at the start of the second row, as distinguished by writing
+// different values after the 132nd. We check multiple rows to be sure.
+// Wrapping at 128 infers CASET XE[7:0]=0x7F, which infers GM[2:0]='011'.
+// Wrapping at 132 infers CASET XE[7:0]=0x83, which infers GM[2:0]='000'.
+//
+// GM[2:0]='000' (132x162) is incorrect for the 128x160 panel actually present,
+// meaning minor coordinate offsets are needed. The offsets are 2 pixels
+// in the narrow dimension (x if portrait) and 1 pixel in the wide dimension
+// (y if portrait). This is due to how the ST7735 controller maps the
+// contents of the internal frame buffer to display itself. See the
+// (unfortunately error-ridden) ST7735 datasheet for more details.
+//
+// NOTE1: Must be run after a HW or SW reset and before any CASET commands.
+// NOTE2: Does NOT always perform a reset before returning. State may be dirty.
+bool need_offset(Capability<volatile OpenTitanUart> &uart, LCD_Interface *interface) {
```

Comment:
The interface here may be passed as a pointer to const (`const LCD_Interface *`).